csharp/src/Ice/Reference.cs

```diff
@@ -1430,194 +1239,83 @@ private void ClearConnection(Connection connection)
                 }
+            }
 
-                if (endpoints.Count == 0)
+            // Apply overrides and filter endpoints
+            IEnumerable<Endpoint> filteredEndpoints = endpoints.Where(endpoint =>
+            {
+                // Filter out opaque and universal endpoints
+                if (endpoint is OpaqueEndpoint || endpoint is UniversalEndpoint)
                 {
-                    throw new NoEndpointException(ToString());
+                    return false;
                 }
 
-                // Apply overrides and filter endpoints
-                IEnumerable<Endpoint> filteredEndpoints = endpoints.Where(endpoint =>
-                {
-                    // Filter out opaque endpoints
-                    if (endpoint is OpaqueEndpoint || endpoint is UniversalEndpoint)
-                    {
-                        return false;
-                    }
-
-                    // Filter out based on InvocationMode and IsDatagram
-                    switch (InvocationMode)
-                    {
-                        case InvocationMode.Twoway:
-                        case InvocationMode.Oneway:
-                        case InvocationMode.BatchOneway:
-                            if (endpoint.IsDatagram)
-                            {
-                                return false;
-                            }
-                            break;
-
-                        case InvocationMode.Datagram:
-                        case InvocationMode.BatchDatagram:
-                            if (!endpoint.IsDatagram)
-                            {
-                                return false;
-                            }
-                            break;
-
-                        default:
-                            Debug.Assert(false);
-                            return false;
-                    }
-
-                    // If PreferNonSecure is false, filter out all non-secure endpoints
-                    return PreferNonSecure || endpoint.IsSecure;
-                });
-
-                if (PreferNonSecure)
+                // Check if the endpoint is compatible with the proxy invocation mode, Twoway and Oneway invocation
+                // modes require a non datagram endpoint, Datagram invocation mode requires a datagram endpoint, the
+                // other invocation modes (BatchOneway and BatchDagram) are not supported.
+                if (InvocationMode switch
                 {
-                    // It's just a preference: we can fallback to secure endpoints.
-                    filteredEndpoints = filteredEndpoints.OrderBy(endpoint => endpoint.IsSecure);
-                }
-
-                endpoints = filteredEndpoints.ToArray();
-                if (endpoints.Count == 0)
+                    InvocationMode.Twoway or InvocationMode.Oneway => endpoint.IsDatagram,
+                    InvocationMode.Datagram => !endpoint.IsDatagram,
+                    _ => true
+                })
```

Comment:
This closes the if for the switch expression. The comment should read, check if the endpoint is NOT compatible ...
```
if (InvocationMode switch
{
    InvocationMode.Twoway or InvocationMode.Oneway => endpoint.IsDatagram,
    InvocationMode.Datagram => !endpoint.IsDatagram,
    _ => true
})
{
    return false;
}
```

When the endpoint is not compatible with the invocation mode we enter the if and `return false` from the closure filters out the endpoint.

Reply:
I don't find this code readable. Why not a simpler:
```
return InvocationMode switch
{
    InvocationMode.Twoway or InvocationMode.Oneway => !endpoint.IsDatagram,
    InvocationMode.Datagram => endpoint.IsDatagram,
    _ => false
}
```
?

```

Reply:
We can if we put this last, we don't want to return `true` here without doing the check after this

```
// If PreferNonSecure is false, filter out all non-secure endpoints
return PreferNonSecure || endpoint.IsSecure;
```

Reply:
```
// If PreferNonSecure is false, filter out all non-secure endpoints
if (!PreferNonSecure && !endpoint.IsSecure)
{
    return false;
}

// Check if the endpoint is compatible with the proxy invocation mode, Twoway and Oneway invocation
// modes require a non datagram endpoint, Datagram invocation mode requires a datagram endpoint, the
// other invocation modes (BatchOneway and BatchDagram) are not supported.
return InvocationMode switch
{
    InvocationMode.Twoway or InvocationMode.Oneway => !endpoint.IsDatagram,
    InvocationMode.Datagram => endpoint.IsDatagram,
    _ => false
};
```

Reply:
Sounds fine, except it will become with Joe's PR:
```
// With ice1, an endpoint is either always secure or always non-secure, so we filter out all non-secure endpoints when PreferNonSecure is false.
if (Protocol == Protocol.Ice1 && !PreferNonSecure && !endpoint.IsAlwaysSecure)
{
    return false;
}
```

Reply:
Right, we have a few conflicts to solve between this PR and Joe's PR